File: app/src/main/java/com/worldventures/dreamtrips/modules/trips/presenter/FiltersPresenter.java
package com.worldventures.dreamtrips.modules.trips.presenter;

import android.os.Bundle;

import com.innahema.collections.query.queriables.Queryable;
import com.worldventures.dreamtrips.core.repository.SnappyRepository;
import com.worldventures.dreamtrips.core.utils.tracksystem.TrackingHelper;
import com.worldventures.dreamtrips.modules.common.api.janet.command.TripsFilterDataCommand;
import com.worldventures.dreamtrips.modules.common.delegate.QueryTripsFilterDataInteractor;
import com.worldventures.dreamtrips.modules.common.presenter.Presenter;
import com.worldventures.dreamtrips.modules.trips.delegate.ResetFilterEventDelegate;
import com.worldventures.dreamtrips.modules.trips.delegate.TripFilterEventDelegate;
import com.worldventures.dreamtrips.modules.trips.model.ActivityModel;
import com.worldventures.dreamtrips.modules.trips.model.RegionModel;
import com.worldventures.dreamtrips.modules.trips.model.TripsFilterDataAnalyticsWrapper;
import com.worldventures.dreamtrips.modules.trips.model.filter.DateFilterItem;
import com.worldventures.dreamtrips.modules.trips.model.filter.FilterFavoriteModel;
import com.worldventures.dreamtrips.modules.trips.model.filter.FilterModel;
import com.worldventures.dreamtrips.modules.trips.model.filter.FilterRecentlyAddedModel;
import com.worldventures.dreamtrips.modules.trips.model.filter.FilterSoldOutModel;
import com.worldventures.dreamtrips.modules.trips.model.filter.RegionHeaderModel;
import com.worldventures.dreamtrips.modules.trips.model.filter.ThemeHeaderModel;
import com.worldventures.dreamtrips.util.TripsFilterData;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import icepick.State;
import io.techery.janet.helper.ActionStateSubscriber;

public class FiltersPresenter extends Presenter<FiltersPresenter.View> {

   @Inject SnappyRepository db;
   @Inject TripFilterEventDelegate tripFilterEventDelegate;
   @Inject ResetFilterEventDelegate resetFilterEventDelegate;
   @Inject QueryTripsFilterDataInteractor queryTripsFilterDataInteractor;

   @State FilterModel filterModel;
   @State ThemeHeaderModel themeHeaderModel;
   @State RegionHeaderModel regionHeaderModel;
   @State FilterSoldOutModel soldOutModel;
   @State FilterRecentlyAddedModel recentlyAddedModel;
   @State FilterFavoriteModel favoriteModel;
   @State DateFilterItem dateFilterItem;
   @State TripsFilterData tripFilterData;

   @Override
   public void restoreInstanceState(Bundle savedState) {
      super.restoreInstanceState(savedState);
      if (savedState == null) {
         filterModel = new FilterModel();
         dateFilterItem = new DateFilterItem();
         themeHeaderModel = new ThemeHeaderModel();
         soldOutModel = new FilterSoldOutModel();
         favoriteModel = new FilterFavoriteModel();
         recentlyAddedModel = new FilterRecentlyAddedModel();
         regionHeaderModel = new RegionHeaderModel();
      }
   }

   @Override
   public void takeView(View view) {
      super.takeView(view);
      subscribeToFiltersLoading();
      if (tripFilterData == null) {
         tripFilterData = new TripsFilterData();
      }
      loadFilters();
      observeResetFilters();
   }

   private void subscribeToFiltersLoading() {
      queryTripsFilterDataInteractor.pipe()
            .observe()
            .compose(bindViewToMainComposer())
            .subscribe(new ActionStateSubscriber<TripsFilterDataCommand>().onSuccess(tripsFilterDataCommand -> {
               view.hideProgress();
               tripFilterData.setAllParentActivities(tripsFilterDataCommand.getParentActivities());
               tripFilterData.setAllRegions(tripsFilterDataCommand.getRegions());
               fillData();
            }).onFail((tripsFilterDataCommand, throwable) -> {
               view.hideProgress();
               view.showErrorContainer();
            }));
   }

   private void observeResetFilters() {
      resetFilterEventDelegate.getObservable()
            .compose(bindViewToMainComposer())
            .subscribe(object -> resetFilters());
   }

   public void loadFilters() {
      view.showProgress();
      queryTripsFilterDataInteractor.pipe().send(new TripsFilterDataCommand());
   }

   public void fillData() {
      List<Object> data = new ArrayList<>();
      data.clear();
      data.add(dateFilterItem);
      data.add(filterModel);
      data.add(soldOutModel);
      data.add(favoriteModel);
      data.add(recentlyAddedModel);

      data.add(regionHeaderModel);
      if (!regionHeaderModel.isHide()) {
         data.addAll(tripFilterData.getAllRegions());
      }

      data.add(themeHeaderModel);

      if (!themeHeaderModel.isHide()) {
         data.addAll(tripFilterData.getAllParentActivities());
      }

      view.fillData(data);
   }

   public void setRegionsChecked(boolean isChecked) {
      for (RegionModel region : tripFilterData.getAllRegions()) {
         region.setChecked(isChecked);
      }
   }

   public void setThemesChecked(boolean isChecked) {
      for (ActivityModel activity : tripFilterData.getAllParentActivities()) {
         activity.setChecked(isChecked);
      }
   }

   public void acceptFilters() {
      tripFilterEventDelegate.post(tripFilterData);
      TrackingHelper.actionFilterTrips(new TripsFilterDataAnalyticsWrapper(tripFilterData));
   }

   public void resetFilters() {
      dateFilterItem.reset();
      filterModel.reset();
      themeHeaderModel.setChecked(true);
      regionHeaderModel.setChecked(true);
      soldOutModel.setActive(false);
      favoriteModel.setActive(false);
      recentlyAddedModel.setActive(false);
      setRegionsChecked(true);
      setThemesChecked(true);
      view.dataSetChanged();
      //
      tripFilterData = new TripsFilterData();
      tripFilterEventDelegate.post(tripFilterData);
      TrackingHelper.actionFilterTrips(new TripsFilterDataAnalyticsWrapper(tripFilterData));
   }

   public void onRangeBarDurationEvent(int minNights, int maxNights) {
      tripFilterData.setMinNights(minNights);
      tripFilterData.setMaxNights(maxNights);
   }

   public void onRangeBarPriceEvent(double minPrice, double maxPrice) {
      tripFilterData.setMinPrice(minPrice);
      tripFilterData.setMaxPrice(maxPrice);
   }

   public void toggleThemeVisibility() {
      themeHeaderModel.setHide(!themeHeaderModel.isHide());
      fillData();
   }

   public void toggleRegionVisibility() {
      regionHeaderModel.setHide(!regionHeaderModel.isHide());
      fillData();
   }

   public void onFilterShowSoldOutEvent(boolean isSoldOut) {
      tripFilterData.setShowSoldOut(isSoldOut);
   }

   public void onFilterShowFavoritesEvent(boolean isSoldOut) {
      tripFilterData.setShowFavorites(isSoldOut);
   }

   public void onFilterShowRecentlyAddedEvent(boolean enabled) {
      tripFilterData.setShowRecentlyAdded(enabled);
   }

   public void onCheckBoxAllRegionsPressedEvent(boolean isChecked) {
      setRegionsChecked(isChecked);
      view.dataSetChanged();
   }

   public void onCheckBoxAllThemePressedEvent(boolean isChecked) {
      setThemesChecked(isChecked);
      view.dataSetChanged();
   }

   public void onThemeSetChangedEvent() {
      boolean allIsChecked = true;
      for (ActivityModel activity : tripFilterData.getAllParentActivities()) {
         if (!activity.isChecked()) {
            allIsChecked = false;
            break;
         }
      }
      themeHeaderModel.setChecked(allIsChecked);
      view.dataSetChanged();
   }

   public void onRegionSetChangedEvent() {
      boolean allIsChecked = true;
      for (RegionModel region : tripFilterData.getAllRegions()) {
         if (!region.isChecked()) {
            allIsChecked = false;
            break;
         }
      }
      regionHeaderModel.setChecked(allIsChecked);
      view.dataSetChanged();
   }

   public void onDatesChanged(DateFilterItem item) {
      tripFilterData.setStartDate(item.getStartDate());
      tripFilterData.setEndDate(item.getEndDate());
   }

   public interface View extends Presenter.View {

      void showProgress();

      void hideProgress();

      void showErrorContainer();

      void hideErrorContainer();

      void fillData(List data);

      void dataSetChanged();
   }
}

File: app/src/main/java/com/worldventures/dreamtrips/modules/trips/presenter/FiltersPresenter.kt
package com.worldventures.dreamtrips.modules.trips.presenter

import com.worldventures.dreamtrips.modules.common.presenter.Presenter
import com.worldventures.dreamtrips.modules.trips.model.analytics.TripsFilterDataAnalyticsWrapper
import com.worldventures.dreamtrips.modules.trips.model.filter.DateFilterItem
import com.worldventures.dreamtrips.modules.trips.model.filter.FilterFavoriteModel
import com.worldventures.dreamtrips.modules.trips.model.filter.FilterModel
import com.worldventures.dreamtrips.modules.trips.model.filter.FilterRecentlyAddedModel
import com.worldventures.dreamtrips.modules.trips.model.filter.FilterSoldOutModel
import com.worldventures.dreamtrips.modules.trips.model.filter.RegionHeaderModel
import com.worldventures.dreamtrips.modules.trips.model.filter.ThemeHeaderModel
import com.worldventures.dreamtrips.modules.trips.model.filter.TripsFilterData
import com.worldventures.dreamtrips.modules.trips.service.TripsInteractor
import com.worldventures.dreamtrips.modules.trips.service.analytics.AcceptDreamTripsFiltersAnalyticAction
import com.worldventures.dreamtrips.modules.trips.service.command.GetTripsFilterDataCommand
import com.worldventures.dreamtrips.modules.trips.service.command.TripFiltersAppliedCommand
import icepick.State
import io.techery.janet.helper.ActionStateSubscriber
import java.util.Date
import javax.inject.Inject

class FiltersPresenter : Presenter<FiltersPresenter.View>() {

   @Inject lateinit var tripsInteractor: TripsInteractor

   @JvmField @State internal var tripFilterData: TripsFilterData? = null
   @JvmField @State internal var regionsExpanded = false
   @JvmField @State internal var regionsSelected = true
   @JvmField @State internal var themesExpanded = false
   @JvmField @State internal var themesSelected = true

   override fun takeView(view: View) {
      super.takeView(view)
      if (tripFilterData == null) tripFilterData = TripsFilterData()
      subscribeToFiltersLoading()
      loadFilters()
   }

   private fun subscribeToFiltersLoading() {
      tripsInteractor.tripFiltersPipe
            .observe()
            .compose(bindViewToMainComposer())
            .subscribe(ActionStateSubscriber<GetTripsFilterDataCommand>()
                  .onStart { view.showProgress() }
                  .onSuccess {
                     tripFilterData?.let { data ->
                        it.result?.let { cachedFilters ->
                           data.allParentActivities = cachedFilters.activities.filter { it.isParent }
                           data.allRegions = cachedFilters.regions
                        }
                     }
                     fillData()
                  }
                  .onFinish { view.hideProgress() }
                  .onFail { _, _ -> view.showErrorContainer() })
   }

   fun loadFilters() = tripsInteractor.tripFiltersPipe.send(GetTripsFilterDataCommand())

   private fun fillData() {
      tripFilterData?.let {
         view.fillData(mutableListOf<Any>().apply {
            add(DateFilterItem(it.startDate, it.endDate))
            add(FilterModel(it))
            add(FilterSoldOutModel(it.isShowSoldOut))
            add(FilterFavoriteModel(it.isShowFavorites))
            add(FilterRecentlyAddedModel(it.isShowRecentlyAdded))
            add(RegionHeaderModel(regionsSelected))
            if (regionsExpanded) addAll(it.allRegions)
            add(ThemeHeaderModel(themesSelected))
            if (themesExpanded) addAll(it.allParentActivities)
         })
      }
   }

   private fun setRegionsChecked(isChecked: Boolean) {
      tripFilterData?.let { for (region in it.allRegions) region.isChecked = isChecked }
   }

   private fun setThemesChecked(isChecked: Boolean) {
      tripFilterData?.let { for (activity in it.allParentActivities) activity.isChecked = isChecked }
   }

   fun acceptFilters() {
      tripFilterData?.let { tripsInteractor.tripFiltersAppliedPipe.send(TripFiltersAppliedCommand(it)) }
      trackApplyFilter()
   }

   fun resetFilters() {
      themesSelected = true
      regionsSelected = true
      setRegionsChecked(true)
      setThemesChecked(true)
      tripFilterData?.reset()
      acceptFilters()
      fillData()
   }

   private fun trackApplyFilter() {
      tripFilterData?.let {
         analyticsInteractor.analyticsActionPipe().send(AcceptDreamTripsFiltersAnalyticAction(TripsFilterDataAnalyticsWrapper(it)))
      }
   }

   fun onRangeBarDurationEvent(minNights: Int, maxNights: Int) {
      tripFilterData?.let {
         it.minNights = minNights
         it.maxNights = maxNights
      }
   }

   fun onRangeBarPriceEvent(minPrice: Double, maxPrice: Double) {
      tripFilterData?.let {
         it.minPrice = minPrice
         it.maxPrice = maxPrice
      }
   }

   fun toggleThemeVisibility() {
      themesExpanded = !themesExpanded
      fillData()
   }

   fun toggleRegionVisibility() {
      regionsExpanded = !regionsExpanded
      fillData()
   }

   fun onFilterShowSoldOutEvent(enabled: Boolean) {
      tripFilterData?.isShowSoldOut = enabled
   }

   fun onFilterShowFavoritesEvent(enabled: Boolean) {
      tripFilterData?.isShowFavorites = enabled
   }

   fun onFilterShowRecentlyAddedEvent(enabled: Boolean) {
      tripFilterData?.isShowRecentlyAdded = enabled
   }

   fun onCheckBoxAllRegionsPressedEvent(isChecked: Boolean) {
      regionsSelected = isChecked
      setRegionsChecked(isChecked)
      fillData()
   }

   fun onCheckBoxAllThemePressedEvent(isChecked: Boolean) {
      themesSelected = isChecked
      setThemesChecked(isChecked)
      fillData()
   }

   fun onThemeSetChangedEvent() {
      themesSelected = true
      tripFilterData?.let {
         for (activity in it.allParentActivities) {
            if (!activity.isChecked) {
               themesSelected = false
               break
            }
         }
      }
      fillData()
   }

   fun onRegionSetChangedEvent() {
      regionsSelected = true
      tripFilterData?.let {
         for ((_, _, isChecked) in it.allRegions) {
            if (!isChecked) {
               regionsSelected = false
               break
            }
         }
      }
      fillData()
   }

   fun onStartDateChanged(date: Date) {
      tripFilterData?.startDate = date
   }

   fun onEndDateChanged(date: Date) {
      tripFilterData?.endDate = date
   }

   interface View : Presenter.View {
      fun showProgress()

      fun hideProgress()

      fun showErrorContainer()

      fun hideErrorContainer()

      fun fillData(data: List<*>)

      fun dataSetChanged()
   }
}
